<a href="https://colab.research.google.com/github/VisarBuza/BachelorThesis_StreamingVideos/blob/main/phase1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

In this notebook we preprocess a data set of Amazon Prime Movies and TV shows.

The dataset can be found [here](https://www.kaggle.com/shivamb/amazon-prime-movies-and-tv-shows).

In [531]:
import pandas as pd 
import collections
import plotly.graph_objs as go


df = pd.read_csv("/content/sample_data/amazon_prime_titles.csv")

## Dataset exploration and modification

### Data types
The data set has 12 columns.

**type** and **rating** are categorical data types.

**release_year** is a numerical data type.

**show_id**, and **rating** are alphanumerical data types

In [532]:
len(df.columns)

12

In [533]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [534]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9668 entries, 0 to 9667
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       9668 non-null   object
 1   type          9668 non-null   object
 2   title         9668 non-null   object
 3   director      7586 non-null   object
 4   cast          8435 non-null   object
 5   country       672 non-null    object
 6   date_added    155 non-null    object
 7   release_year  9668 non-null   int64 
 8   rating        9331 non-null   object
 9   duration      9668 non-null   object
 10  listed_in     9668 non-null   object
 11  description   9668 non-null   object
dtypes: int64(1), object(11)
memory usage: 906.5+ KB


In [535]:
df.nunique()

show_id         9668
type               2
title           9668
director        5774
cast            7927
country           86
date_added        84
release_year     100
rating            24
duration         219
listed_in        518
description     9414
dtype: int64

## Data quality
After looking throught the data we can find several issues with it.

First we can see that the date_added, and country colums are provided only for 155 and 672 rcords respectively. We can go ahead and drop these columns since they provides no additional value.

The director and cast values are missing for some of the records.

In [536]:
df.describe(include = 'all')

show_id   type                  title  ...  duration listed_in description
count     9668   9668                   9668  ...      9668      9668        9668
unique    9668      2                   9668  ...       219       518        9414
top      s7010  Movie  Call Time: The Finale  ...  1 Season     Drama           1
freq         1   7814                      1  ...      1342       986          15
mean       NaN    NaN                    NaN  ...       NaN       NaN         NaN
std        NaN    NaN                    NaN  ...       NaN       NaN         NaN
min        NaN    NaN                    NaN  ...       NaN       NaN         NaN
25%        NaN    NaN                    NaN  ...       NaN       NaN         NaN
50%        NaN    NaN                    NaN  ...       NaN       NaN         NaN
75%        NaN    NaN                    NaN  ...       NaN       NaN         NaN
max        NaN    NaN                    NaN  ...       NaN       NaN         NaN

[11 rows x 12 columns]

We cal also check if the data set has duplicate rows, but after checking we find that the dataset has no duplicate rows.

In [537]:
duplicate = df[df.duplicated()]
duplicate.head()

Empty DataFrame
Columns: [show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description]
Index: []

In [538]:
df.loc[df['director'] == '1']

show_id   type  ...  listed_in                            description
6930   s6931  Movie  ...     Action                Title Post onboarding 8
7127   s7128  Movie  ...     Action                                      1
7128   s7129  Movie  ...     Action                                      1
7139   s7140  Movie  ...     Action                                  clip1
7238   s7239  Movie  ...     Action    ACT 4 - Title after removing 1C - 5
7239   s7240  Movie  ...     Action  ACT 4 - Title after removing 1C - 2 a
7245   s7246  Movie  ...     Action      ACT 4 - Title after removing 1C -
7258   s7259  Movie  ...     Comedy   ACT 4 - Title after removing 1C - 44
8501   s8502  Movie  ...     Action                        Act 6 - Title 1
8502   s8503  Movie  ...     Action                        Act 5 - Title 1
8755   s8756  Movie  ...     Action                                      1
8756   s8757  Movie  ...     Action                                      a
8757   s8758  Movie  ...     Action                                      1
8758   s8759  Movie  ...     Action                                     1a
9462   s9463  Movie  ...  Animation                                      1
9463   s9464  Movie  ...     Action                                      1

[16 rows x 12 columns]

Some of the values for director have values of 1, so we can go ahead and drop those records since they bring no value.

In [539]:
df = df[df.director != '1']

In [540]:
df.loc[df['rating'] == 'AGES_18_']


show_id  ...                                        description
5132   s5133  ...  100 Streets is a powerful ensemble drama, foll...
5150   s5151  ...  Francesca Eastwood gives a breakout performanc...
5734   s5735  ...  Ucult Studios presents a woman realizes that b...

[3 rows x 12 columns]

Some of the values for rating can be normalized as follows:

NOT_RATE and NR can be mapped to UNRATED

ALL_AGES can be mapped to ALL

AGES_16_ can be mapped to 16+

AGES_18_ can be mapped to 18+

In [541]:
df['rating'] = df['rating'].fillna('ALL')
df['rating'] = df.apply(lambda x : 'UNRATED' if "NOT_RATE" in x['rating'] else x['rating'], axis = 1)
df['rating'] = df.apply(lambda x : 'UNRATED' if "NR" in x['rating'] else x['rating'], axis = 1)
df['rating'] = df.apply(lambda x : 'ALL' if "ALL_AGES" in x['rating'] else x['rating'], axis = 1)
df['rating'] = df.apply(lambda x : '16+' if "AGES_16_" in x['rating'] else x['rating'], axis = 1)
df['rating'] = df.apply(lambda x : '18+' if "AGES_18_" in x['rating'] else x['rating'], axis = 1)

## Missing values

There are a lot of missing values

In [542]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2082
cast            1233
country         8980
date_added      9497
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

In [543]:
df.drop(columns=['date_added', 'country'], inplace = True)

For the rating column, we can estimate that they are applicable to all ages and will mark them as **ALL**

In [544]:
df['rating'] = df['rating'].fillna('ALL')

In [545]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2082
cast            1233
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

In [546]:
df.describe(include="all")

show_id  ...                                        description
count     9652  ...                                               9652
unique    9652  ...                                               9405
top      s7010  ...  Arpo was designed as a commando robot, built a...
freq         1  ...                                                 15
mean       NaN  ...                                                NaN
std        NaN  ...                                                NaN
min        NaN  ...                                                NaN
25%        NaN  ...                                                NaN
50%        NaN  ...                                                NaN
75%        NaN  ...                                                NaN
max        NaN  ...                                                NaN

[11 rows x 10 columns]

## Sampling

Sampling is often used for two reasons:

*   Collecting all the data is too expensive or time consuming
*   Analysing the data takes too much computational power

Point one does not apply here, the data has already been collected. Point two can also be ignored because our data set is relatively small



## Dimensionality reduction

We have columns that are not needed for further processing of the data, so we can go ahead and remove them.


In [547]:
df.drop(columns=['description','show_id'], inplace = True)

## Transformation

The duration column has the values as string, and it differes based on type of the the record.
If the record is a movie, the column of duration accounts for the minutes of the duration.
If the record is a tv show, the column of duration accounts for the number of seasons that the show has.

Tv Show: 1 Season
Movie : 113 min

We can normalize the duration field for movies by removing the min part of the string.
For tv shows we can add a new feature labeled season_count which is the normalized value of how many seasons that tv show has.

In [548]:
df['season_count'] = df.apply(lambda x : x['duration'].split(" ")[0] if "Season" in x['duration'] else "", axis = 1)
df['duration'] = df.apply(lambda x : x['duration'].split(" ")[0] if "Season" not in x['duration'] else "", axis = 1)

## Feature creation

All of the movies or tv shows presented here fall under some genre in the **listed_in** column. These are delimited with **' , '**. New columns can be created to represent a primary genre, secondary genre and so forth.

In [549]:
df[['genre 1', 'genre 2', 'genre 3', 'genre 4', 'genre 5']] = df['listed_in'].str.split(',', expand=True)
df.drop(columns=['listed_in'], inplace=True)

## Binarization

In the dataset we are using, there are only two columns for type, so we can change the name to IsMovie and binarize the values

In [550]:
df.loc[df['type'] == 'Movie', 'type'] = 1
df.loc[df['type'] == 'TV Show', 'type'] = 0
df.rename(columns={'type': 'is_movie'}, inplace = True)
df.head()

is_movie                 title        director  ...    genre 3  genre 4 genre 5
0        1   The Grand Seduction    Don McKellar  ...       None     None    None
1        1  Take Care Good Night    Girish Joshi  ...       None     None    None
2        1  Secrets of Deception     Josh Webber  ...   Suspense     None    None
3        1    Pink: Staying True  Sonia Anderson  ...       None     None    None
4        1         Monster Maker    Giles Foster  ...       None     None    None

[5 rows x 13 columns]

Here is a sample of the preprocessed data set

In [551]:
df.sample(n = 30)

is_movie  ...            genre 5
9014        0  ...               None
1916        0  ...               None
8107        1  ...               None
2316        1  ...               None
2660        1  ...               None
7316        1  ...               None
8500        1  ...               None
9215        1  ...               None
7814        1  ...               None
5981        1  ...               None
9473        1  ...               None
7767        1  ...               None
8940        1  ...               None
7390        1  ...               None
8505        0  ...               None
7250        0  ...               None
3006        1  ...               None
8555        1  ...               None
4524        1  ...               None
166         1  ...               None
7956        1  ...               None
316         1  ...   Special Interest
702         1  ...               None
1485        1  ...               None
9593        1  ...               None
1709        1  ...               None
3766        1  ...               None
5481        1  ...               None
9003        1  ...               None
8474        1  ...               None

[30 rows x 13 columns]

## Insights
In the next section we provided some insights about how the dataset might be useful

TV shows

In [552]:
tvshows = df[df['is_movie'] == 0]
tvshows.drop(columns=['director'], inplace = True)
tvshows.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1854 entries, 17 to 9666
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   is_movie      1854 non-null   object
 1   title         1854 non-null   object
 2   cast          1385 non-null   object
 3   release_year  1854 non-null   int64 
 4   rating        1854 non-null   object
 5   duration      1854 non-null   object
 6   season_count  1854 non-null   object
 7   genre 1       1854 non-null   object
 8   genre 2       1060 non-null   object
 9   genre 3       376 non-null    object
 10  genre 4       27 non-null     object
 11  genre 5       12 non-null     object
dtypes: int64(1), object(11)
memory usage: 188.3+ KB


Movies

In [553]:
movies = df[df['is_movie'] == 1]
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7798 entries, 0 to 9667
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   is_movie      7798 non-null   object
 1   title         7798 non-null   object
 2   director      7570 non-null   object
 3   cast          7034 non-null   object
 4   release_year  7798 non-null   int64 
 5   rating        7798 non-null   object
 6   duration      7798 non-null   object
 7   season_count  7798 non-null   object
 8   genre 1       7798 non-null   object
 9   genre 2       4769 non-null   object
 10  genre 3       1680 non-null   object
 11  genre 4       406 non-null    object
 12  genre 5       311 non-null    object
dtypes: int64(1), object(12)
memory usage: 852.9+ KB
